### PR TITLE
#1244: set teams lead

### LINKED ISSUE #1244: Teams - Set Team Leads Endpoint
### Description

Now that teams have leads as well as heads, we need to add the ability to set the leads of a team. For this ticket we wanna add the backend endpoint that allows a user to set the leads of a team.

### Acceptance Criteria

- The new route should be in the teams router and be '/:teamId/set-leads' and take an array of user ids in it's payload
- Leads cannot be members or the head of the team they're being added to.
- Leads must not be a head or a lead on another existing team
- Only an admin or a head of the team can update the team's leads

### Proposed Solution

Add this new endpoint to the teams controller, service method, router, etc.

### Mocks

epic #1255

### PR DESCRIPTION
**Pull request by ryanhowe28**
## Changes

Made an endpoint to be able to edit a teams lead
## Notes

I'm back
Tested in postman
Before
image
Post
image
image
image
image

## Checklist

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #1244

### CODE REVIEW

File: src/backend/src/routes/teams.routes.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ teamsRouter.post(
   validateInputs,
   TeamsController.setTeamMembers
 );
+teamsRouter.post('/:teamId/set-leads', intMinZero(body('userId')), validateInputs, TeamsController.setTeamLead);
```
Comment (Peyton-McKee) [inline src/backend/src/routes/teams.routes.ts:18]: this is an array of userIds

File: src/backend/src/services/teams.services.ts [hunk 2]
```diff
@@ -162,4 +162,45 @@ export default class TeamsService {
 
     return teamTransformer(updateTeam);
   }
+
+  static async setTeamLead(submitter: User, teamId: string, userId: number): Promise<Team> {
```
Comment (Peyton-McKee) [inline src/backend/src/services/teams.services.ts:166]: setTeamLeads its many leads will be an array of userIds + javadoc

File: src/backend/src/services/teams.services.ts [hunk 3]
```diff
@@ -162,4 +162,45 @@ export default class TeamsService {
...
+    });
+
+    if (!team) throw new NotFoundException('Team', teamId);
+    if (!newLead) throw new NotFoundException('User', userId);
+
+    if (!isAdmin(submitter.role) && submitter.userId !== team.headId) {
+      throw new AccessDeniedException('You must be an admin or the head to update the lead!');
+    }
+
+    if (newLead.userId === team.headId || team.members.includes(newLead)) {
```
Comment (Peyton-McKee) [inline src/backend/src/services/teams.services.ts:183]: there should be a function for this in shared called like makeTeamList aslso these will all be lists not just one person

File: src/backend/src/services/teams.services.ts [hunk 4]
```diff
@@ -162,4 +162,45 @@ export default class TeamsService {
...
     return teamTransformer(updateTeam);
   }
+
+  static async setTeamLead(submitter: User, teamId: string, userId: number): Promise<Team> {
+    const team = await prisma.team.findUnique({
+      where: { teamId },
+      ...teamQueryArgs
+    });
+
+    const newLead = await prisma.user.findUnique({
```
Comment (Peyton-McKee) [inline src/backend/src/services/teams.services.ts:172]: gonna be a find many

File: src/backend/src/services/teams.services.ts [hunk 5]
```diff
@@ -162,4 +162,45 @@ export default class TeamsService {
...
+    if (!isAdmin(submitter.role) && submitter.userId !== team.headId) {
+      throw new AccessDeniedException('You must be an admin or the head to update the lead!');
+    }
+
+    if (newLead.userId === team.headId || team.members.includes(newLead)) {
+      throw new AccessDeniedException('The lead cannot be the head or a member of the team!');
+    }
+
+    // checking to see if any other teams have the new lead as their current head or lead
+    const newLeadTeam = await prisma.team.findFirst({
```
Comment (Peyton-McKee) [inline src/backend/src/services/teams.services.ts:188]: this is gonna have to be updated

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
looks good. You also need unit tests

**Comment by RChandler234** [comment]
Look at setTeamMembers, this will be almost identical to that

File: src/backend/src/controllers/teams.controllers.ts [hunk 6]
```diff
@@ -62,4 +62,16 @@ export default class TeamsController {
       next(error);
     }
   }
+
+  static async setTeamLead(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { userId } = req.body;
```
Comment (callmejim1226) [inline src/backend/src/controllers/teams.controllers.ts:68]: Nice work overall on setting the controller.

We want to update this to be userIds (array of user ids) after the edit in the router file.

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
Great start!

Have same comments as Peyton, where we wanna make sure to have array of user ids to set team leads. 

Also, can you add screenshots of postman (ex. successful POST request & result and GET request on team to show the team leads got updated), so we can know the endpoint is working fine.

File: src/backend/src/services/teams.services.ts [hunk 7]
```diff
@@ -162,4 +162,59 @@ export default class TeamsService {
...
     return teamTransformer(updateTeam);
   }
+
+  /**
+   * Update the given teamId's team's leads
+   * @param submitter a user who's making this request
+   * @param teamId a id of team to be updated
+   * @param userIds a array of user Ids that replaces team's old leads
+   * @returns an updated team
+   * @throws if the team is not found, the submitter has no priviledge, or any user from the given userIds does not exist
```
Comment (RChandler234) [inline src/backend/src/services/teams.services.ts:172]: privilege*

File: src/backend/src/services/teams.services.ts [hunk 8]
```diff
@@ -162,4 +162,59 @@ export default class TeamsService {
...
+    if (newLeads.map((lead) => lead.userId).includes(team.headId)) {
+      throw new AccessDeniedException('The lead cannot be the head of the team!');
+    }
+
+    if (team.members.map((member) => member.userId).some((memberId) => userIds.includes(memberId))) {
+      throw new AccessDeniedException('The lead cannot be a member of the team!');
+    }
+
+    if (teams.map((team) => team.headId).some((teamHeadId) => newLeads.map((lead) => lead.userId).includes(teamHeadId))) {
+      throw new AccessDeniedException('A teams lead must not be a head of another team!');
```
Comment (RChandler234) [inline src/backend/src/services/teams.services.ts:199]: we should also verify that none of the new leads are a lead on another team

File: src/backend/src/services/teams.services.ts [hunk 9]
```diff
@@ -162,4 +162,59 @@ export default class TeamsService {
...
+
+    const newLeads = await getUsers(userIds);
+
+    if (!team) throw new NotFoundException('Team', teamId);
+
+    if (!isAdmin(submitter.role) && submitter.userId !== team.headId) {
+      throw new AccessDeniedException('You must be an admin or the head to update the lead!');
+    }
+
+    if (newLeads.map((lead) => lead.userId).includes(team.headId)) {
```
Comment (RChandler234) [inline src/backend/src/services/teams.services.ts:190]: these bottom 3 should be HttpExceptions with the 400 error code, we should also add tests for each of these

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/services/teams.services.ts (hunk 9 above)
Comment (callmejim1226) [reply src/backend/src/services/teams.services.ts:190]: To add on, we should also add a test that validates if none of the new leads are a lead on another team.

Comment (callmejim1226) [reply src/backend/src/services/teams.services.ts:190]: Can we also make a separate variable for `newLeads.map((lead) => lead.userId)` into newLeadIds? This variable will be used multiple times for validating any HttpExceptions, and using it will make us read the code easier, especially on line 198 where we have to validate if Leads is not a head or a lead on another existing team.

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
Great work!

Like the screenshots with the description. 

I left a few comments on unit tests and code refactoring stuff.

File: src/backend/src/services/teams.services.ts (hunk 9 above)
Comment (Peyton-McKee) [reply src/backend/src/services/teams.services.ts:190]: I dont think we should limit leads to one team

File: src/backend/src/services/teams.services.ts (hunk 8 above)
Comment (Peyton-McKee) [reply src/backend/src/services/teams.services.ts:199]: I disagreee

**Comment by RChandler234** [comment]
Backend looks good

File: src/backend/src/services/teams.services.ts [hunk 10]
```diff
@@ -162,4 +162,60 @@ export default class TeamsService {
...
+
+    const newLeads = await getUsers(userIds);
+
+    if (!team) throw new NotFoundException('Team', teamId);
+
+    if (!isAdmin(submitter.role) && submitter.userId !== team.headId) {
+      throw new AccessDeniedException('You must be an admin or the head to update the lead!');
+    }
+
+    console.log(teams);
```
Comment (Peyton-McKee) [inline src/backend/src/services/teams.services.ts:190]: delete

File: src/backend/src/services/teams.services.ts [hunk 11]
```diff
@@ -162,4 +162,60 @@ export default class TeamsService {
...
+   * @returns an updated team
+   * @throws if the team is not found, the submitter has no privilege, or any user from the given userIds does not exist
+   */
+  static async setTeamLeads(submitter: User, teamId: string, userIds: number[]): Promise<Team> {
+    const team = await prisma.team.findUnique({
+      where: { teamId },
+      ...teamQueryArgs
+    });
+
+    const teams = await prisma.team.findMany(teamQueryArgs);
```
Comment (Peyton-McKee) [inline src/backend/src/services/teams.services.ts:180]: so i was refactoring my backend, and I think it makes sense to use our current service functions for these (aka use the getAllTeams service function instead of directly calling the prisma shit)

File: src/backend/src/services/teams.services.ts [hunk 12]
```diff
@@ -162,4 +162,59 @@ export default class TeamsService {
...
+    if (newLeads.map((lead) => lead.userId).includes(team.headId)) {
+      throw new HttpException(400, 'A lead cannot be the head of the team!');
+    }
+
+    if (team.members.map((member) => member.userId).some((memberId) => userIds.includes(memberId))) {
+      throw new HttpException(400, 'A lead cannot be a member of the team!');
+    }
+
+    if (teams.map((team) => team.headId).some((teamHeadId) => newLeads.map((lead) => lead.userId).includes(teamHeadId))) {
+      throw new HttpException(400, 'A teams lead must not be a head of another team!');
```
Comment (RChandler234) [inline src/backend/src/services/teams.services.ts:199]: After a convo with Sofia and Walter, we do actually wanna allow this, so you can remove this check/test. Can you update the head endpoint as well to allow a head to also be a lead on other teams.

**Review by RChandler234 — CHANGES_REQUESTED** [review]
Will do a more thorough look thru in a bit

File: src/frontend/src/apis/teams.api.ts [hunk 13]
```diff
@@ -36,3 +36,9 @@ export const setTeamHead = (id: string, userId: number) => {
     userId
   });
 };
+
+export const setTeamLeads = (id: string, userIds: number[]) => {
+  return axios.post<{ message: string }>(apiUrls.teamsSetLeads(id), {
```
Comment (RChandler234) [inline src/frontend/src/apis/teams.api.ts:41]: this endpoint returns a Team, not an object with a message

File: src/frontend/src/hooks/teams.hooks.ts [hunk 14]
```diff
@@ -67,3 +74,19 @@ export const useEditTeamDescription = (teamId: string) => {
     }
   );
 };
+
+export const useSetTeamLeads = (teamId: string) => {
+  const queryClient = useQueryClient();
+  return useMutation<{ message: string }, Error, any>(
```
Comment (RChandler234) [inline src/frontend/src/hooks/teams.hooks.ts:80]: again here, returns a Team not a message

File: src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx [hunk 15]
```diff
@@ -144,7 +193,14 @@ const TeamMembersPageBlock: React.FC<TeamMembersPageBlockProps> = ({ team }) =>
   );
 
   const NonEditingLeadsView = () => (
-    <DetailDisplay label="Leads" content={team.leads.map((lead) => fullNamePipe(lead)).join(', ')} />
+    <Grid container>
+      <Grid item>
```
Comment (RChandler234) [inline src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx:197]: If you make the screen too small, the edit button goes down to the next line (or if you have too many leads assigned), but we want it to stay on the left. you can fix this by setting the xs property on these 2 items

**Review by RChandler234 — CHANGES_REQUESTED** [review]
Look really good

File: src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx [hunk 16]
```diff
@@ -30,17 +30,21 @@ const TeamMembersPageBlock: React.FC<TeamMembersPageBlockProps> = ({ team }) =>
...
+  const { isLoading: setTeamLeadsIsLoading, mutateAsync: setTeamLeadsMutateAsync } = useSetTeamLeads(team.teamId);
 
   const toast = useToast();
 
   if (allUsersIsError) return <ErrorPage message={allUsersError?.message} />;
-  if (allUsersIsLoading || setTeamMembersIsLoading || setTeamHeadIsLoading || !users) return <LoadingIndicator />;
+  if (allUsersIsLoading || setTeamMembersIsLoading || setTeamHeadIsLoading || setTeamLeadsIsLoading || !users)
+    return <LoadingIndicator />;
 
   const hasPerms = auth.user && (isAdmin(auth.user.role) || auth.user.userId === team.head.userId);
```
Comment (RChandler234) [inline src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx:49]: not on you, but can you update this file to use useCurrentUser rather than useAuth (might have to update some tests as well)

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx [hunk 17]
```diff
@@ -27,22 +26,26 @@ const userToAutocompleteOption = (user: User): { label: string; id: string } =>
 };
 
 const TeamMembersPageBlock: React.FC<TeamMembersPageBlockProps> = ({ team }) => {
-  const auth = useAuth();
+  const auth = useCurrentUser();
```
Comment (RChandler234) [inline src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx:29]: rename to user

File: src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx [hunk 18]
```diff
@@ -27,22 +26,26 @@ const userToAutocompleteOption = (user: User): { label: string; id: string } =>
...
 
   const toast = useToast();
 
   if (allUsersIsError) return <ErrorPage message={allUsersError?.message} />;
-  if (allUsersIsLoading || setTeamMembersIsLoading || setTeamHeadIsLoading || !users) return <LoadingIndicator />;
+  if (allUsersIsLoading || setTeamMembersIsLoading || setTeamHeadIsLoading || setTeamLeadsIsLoading || !users)
+    return <LoadingIndicator />;
 
-  const hasPerms = auth.user && (isAdmin(auth.user.role) || auth.user.userId === team.head.userId);
+  const hasPerms = auth && (isAdmin(auth.role) || auth.userId === team.head.userId);
```
Comment (RChandler234) [inline src/frontend/src/pages/TeamsPage/TeamMembersPageBlock.tsx:48]: don't need to check that it exists since we're using useCurrentUser

**Review by RChandler234 — CHANGES_REQUESTED** [review]

**Review by RChandler234 — APPROVED** [review]
